crates/aptos-telemetry-service/src/custom_telemetry.rs

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use std::time::Duration;
+
+use crate::{
+    auth::with_auth,
+    context::Context,
+    types::auth::{Claims, TelemetryDump},
+};
+use aptos_config::config::PeerRole;
+use aptos_logger::{error, info};
+use gcp_bigquery_client::model::table_data_insert_all_request::TableDataInsertAllRequest;
+use serde_json::json;
+use warp::{filters::BoxedFilter, reject, reply, Filter, Rejection, Reply};
+
+pub fn custom_telemetry(context: Context) -> BoxedFilter<(impl Reply,)> {
+    warp::path!("telemetry")
+        .and(warp::post())
+        .and(context.clone().filter())
+        .and(with_auth(
+            context,
+            vec![PeerRole::Validator, PeerRole::Unknown],
+        ))
+        .and(warp::body::json())
+        .and_then(handle_custom_telemetry)
+        .boxed()
+}
+
+pub async fn handle_custom_telemetry(
+    context: Context,
+    _claims: Claims,
+    body: TelemetryDump,
+) -> anyhow::Result<impl Reply, Rejection> {
+    let mut insert_request = TableDataInsertAllRequest::new();
+
+    let telemetry_event = body.events[0].clone();
```

Comment:
what happens when the events array being passed has 0 items? Will this simply crash?

Reply:
Good catch. Checking for `is_empty` and throwing a refined error.